extensions/opencv/MotionDetector.h

```diff
@@ -0,0 +1,95 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef NIFI_MINIFI_CPP_MOTIONDETECTOR_H
+#define NIFI_MINIFI_CPP_MOTIONDETECTOR_H
+
+#include <atomic>
+
+#include <core/Resource.h>
+#include <core/Processor.h>
+#include <opencv2/opencv.hpp>
+#include <opencv2/objdetect.hpp>
+#include <opencv2/imgproc.hpp>
+#include "FrameIO.h"
+
+#include <iomanip>
+#include <ctime>
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace processors {
+
+class MotionDetector : public core::Processor {
+
+ public:
+
+  explicit MotionDetector(const std::string &name, utils::Identifier uuid = utils::Identifier())
+      : Processor(name, uuid),
+        logger_(logging::LoggerFactory<MotionDetector>::getLogger()) {
+  }
+
+  static core::Property ImageEncoding;
+  static core::Property MinInterestArea;
+  static core::Property Threshold;
+  static core::Property DilateIter;
+  static core::Property BackgroundFrame;
+  static core::Property PrevAsBackground;
+
+  static core::Relationship Success;
+  static core::Relationship Failure;
+
+  void initialize() override;
+  void onSchedule(core::ProcessContext *context, core::ProcessSessionFactory *sessionFactory) override;
+  void onTrigger(core::ProcessContext *context, core::ProcessSession *session) override {
+    logger_->log_error("onTrigger invocation with raw pointers is not implemented");
+  }
+  void onTrigger(const std::shared_ptr<core::ProcessContext> &context,
+                 const std::shared_ptr<core::ProcessSession> &session) override;
+
+  void notifyStop() override;
+
+ private:
+
+  bool detectAndDraw(cv::Mat &frame);
+
+  std::shared_ptr<logging::Logger> logger_;
+  std::mutex mutex_;
+  cv::Mat background_;
+  cv::Mat bg_img_;
+  std::string image_encoding_;
+  int min_area_;
+  int threshold_;
+  int dil_iter_;
+  bool update_bg_;
+
+//   Background & Foreground seperation.
+//  std::vector<cv::Mat> history;
```

Comment:
Are these parts of some features not ready yet or just leftovers?

Reply:
They are for "Background & Foreground extraction" features that I wanted earlier. However, after second thought, I think I will create a new processor for that later, since it can be useful itself.